### ISSUE TITLE
Warn when `core.match` pattern interprets class name as wildcard

### ISSUE THREAD

**Issue by colin-p-hill** [clj-kondo/clj-kondo#1955]
To upvote this issue, give it a thumbs up. See [this list](https://github.com/clj-kondo/clj-kondo/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc) for the most upvoted issues.

**Is your feature request related to a problem? Please describe.**
The `core.match` algorithm has a fairly significant footgun in an edge case, alluded to in its [basic usage documentation](https://github.com/clojure/core.match/wiki/Basic-usage#locals). Without the `let` bindings shown there, the class name symbols will match anything.
```clojure
(match [(type 1)]    ; a java.lang.Class
       [String] :string
       [Long] :long)
;;=> :string
```

**Describe the solution you'd like**
A new linter which issues a finding when a `core.match` pattern contains a symbol that
1. can be resolved to a class and
2. is not locally bound at its usage site.

**Describe alternatives you've considered**
None come to mind.

**Comment by colin-p-hill** [comment]
Come to think of it, maybe this should be expanded to the use of any non-local resolvable symbol?
```
user=> (def my-constant 99)
#'user/my-constant
user=> (match [100] [my-constant] :matched-var :else :no-match)
:matched-var
```
False positives may happen more often that way, though.
